File: packages/ld-adapter/src/resolve/variation-resolver.ts
import type { FlagState, VariationRef } from '@ff-promo/contracts';
import { UnresolvedVariationError } from '../errors/ld-adapter-error.js';

function valuesEqual(a: unknown, b: unknown): boolean {
	return JSON.stringify(a) === JSON.stringify(b);
}

export function resolveVariationId(
	flagState: FlagState,
	ref: VariationRef,
): string {
	const matches = flagState.variations.filter((variation) => {
		switch (ref.by) {
			case 'id':
				return variation.id === ref.id;
			case 'name':
				return variation.name === ref.name;
			case 'value':
				return valuesEqual(variation.value, ref.value);
			default:
				return false;
		}
	});

	if (matches.length !== 1) {
		throw new UnresolvedVariationError(
			`Expected exactly one variation match, found ${matches.length}`,
			{ ref, flagKey: flagState.flagKey, projectKey: flagState.projectKey },
		);
	}

	const match = matches[0];
	if (!match?.id) {
		throw new UnresolvedVariationError('Matched variation is missing id', {
			ref,
			flagKey: flagState.flagKey,
		});
	}

	return match.id;
}

File: kotlin/modules/ld-adapter/src/main/kotlin/com/ffpromo/ldadapter/resolve/VariationResolver.kt
package com.ffpromo.ldadapter.resolve

import com.ffpromo.contracts.FlagState
import com.ffpromo.contracts.VariationRef
import com.ffpromo.ldadapter.errors.UnresolvedVariationError
import kotlinx.serialization.json.Json

private val json = Json { ignoreUnknownKeys = true }

private fun valuesEqual(a: kotlinx.serialization.json.JsonElement, b: kotlinx.serialization.json.JsonElement): Boolean =
    json.encodeToString(kotlinx.serialization.json.JsonElement.serializer(), a) ==
        json.encodeToString(kotlinx.serialization.json.JsonElement.serializer(), b)

fun resolveVariationId(flagState: FlagState, ref: VariationRef): String {
    val matches = flagState.variations.filter { variation ->
        when (ref) {
            is VariationRef.ById -> variation.id == ref.id
            is VariationRef.ByName -> variation.name == ref.name
            is VariationRef.ByValue -> valuesEqual(variation.value, ref.value)
        }
    }

    if (matches.size != 1) {
        throw UnresolvedVariationError(
            message = "Expected exactly one variation match, found ${matches.size}",
            context = mapOf(
                "ref" to ref,
                "flagKey" to flagState.flagKey,
                "projectKey" to flagState.projectKey,
            ),
        )
    }

    val match = matches.first()
    if (match.id.isEmpty()) {
        throw UnresolvedVariationError(
            message = "Matched variation is missing id",
            context = mapOf("ref" to ref, "flagKey" to flagState.flagKey),
        )
    }

    return match.id
}
